plugins/analytics/public/dockerhub.py

```diff
@@ -0,0 +1,201 @@
+        context = {"source": "hub.docker.com"}
+        if metadata:
+            context["Analysis"] = "Image found"
+            context["Image stats"] = {
+                "Pulls": metadata.get("pull_count", "N/A"),
+                "Last updated": metadata.get("last_updated", "N/A"),
+                "Registered": metadata.get("date_registered", "N/A"),
+                "Stars": metadata.get("star_count", "N/A"),
+            }
+            context["Image details"] = metadata.get("tags", "N/A")
+            context["User details"] = metadata.get("user", "Unknown")
+        else:
+            context["Analysis"] = "Image not found"
+        return context
+
+    def _create_digest_context(self, image_obs, tag_name, image_tag):
+        if not image_tag:
+            return {}
+        fullname = f"{image_obs.value}:{tag_name}"
+        arch = image_tag.get("arch", "N/A")
+        os = image_tag.get("os", "N/A")
+        pushed = image_tag.get("last_pushed", "N/A")
+        context = {"image name": fullname, "arch": f"{os}/{arch}", "pushed": pushed}
+        return context
+
+    def _create_digest_observable(self, image_obs, digest, context, link_type):
+        sha_obs = self._get_observable(observable.ObservableType.sha256, value=digest)
+        if context:
+            sha_obs.add_context("hub.docker.com", context)
+            sha_obs.tag({"dockerhub"})
+        image_obs.link_to(sha_obs, link_type, "")
+        return sha_obs
+
+    def _make_relationships(self, image_obs, metadata):
+        username = metadata.get("user", {}).get("username", "")
+        if username:
+            user_obs = self._get_observable(
+                observable.ObservableType.generic, value=username
+            )
+            user_obs.tag({"type:user_account", "dockerhub"})
+            user_obs.link_to(image_obs, "creates", "")
+        for tag_name, image_tags in metadata.get("tags", {}).items():
+            for image_tag in image_tags:
+                context = self._create_digest_context(image_obs, tag_name, image_tag)
+                digest = re.match(
+                    r"sha256:([0-9a-f]{64})$",
+                    image_tag.get("digest", ""),
+                    re.IGNORECASE,
+                )
+                if digest:
+                    self._create_digest_observable(
+                        image_obs, digest.group(1), context, "results_to"
+                    )
+                for layer in image_tag.get("layers", []):
+                    layer_digest = re.match(
+                        r"sha256:([0-9a-f]{64})$",
+                        layer.get("digest", ""),
+                        re.IGNORECASE,
+                    )
+                    if layer_digest:
+                        self._create_digest_observable(
+                            image_obs, layer_digest.group(1), context, "generates"
+                        )
+                    file_digest = re.match(
+                        r"^\w+ file:([0-9a-f]{64}) .*",
```

Comment:
I don't know what volume you're expecting, but consider some optimizations:

* pre-compile the regexes (you're reusing one of them) and declare them at module-level
* instead of passing on a potentially empty string to a match function (`layer.get("instruction", "")`), consider extracting that variable and checking its value before, e.g.

```python
if layer.get("instruction"):
  file_digest = FILE_REGEX.match(layer["instruction"])
  if file_digest:
     self._create_digest_observable()
```

Reply:
Fixed in [063650a](https://github.com/yeti-platform/yeti/pull/904/commits/063650a0fd776d685ae1b20c012e387081a012e2) and [ce4996b](https://github.com/yeti-platform/yeti/pull/904/commits/ce4996bf80995751cfe9ccc88bed9e1006cfd70d)